<a href="https://colab.research.google.com/github/mohamedbahaaeldien92-max/Elezaby-project/blob/main/recovered_failed_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Load "order_products_fact.csv" and "products_dimension.csv" as pandas DataFrames, clean the 'product_id' column in `products_dimension`, and create a reference table using unique 'product_id' and 'unit_price' from 'order_products_fact'.

## Load DataFrames

### Subtask:
Load 'order_products_fact.csv' and 'products_dimension.csv' into pandas DataFrames. Ensure 'product_id' is loaded as a string for both, and 'unit_price' is converted to float. Verify 'product_id' in 'order_products_fact' is entirely of string type.


**Reasoning**:
First, I'll import the pandas library to handle data manipulation. Then, I'll load both CSV files into pandas DataFrames as requested, ensuring the specified data types for 'product_id' and 'unit_price' during loading. Finally, I will verify the 'product_id' column type in 'order_products_df'.



In [1]:
import pandas as pd

# Load 'order_products_fact.csv'
order_products_df = pd.read_csv(
    '/content/order_products_fact.csv',
    dtype={'product_id': str},
    converters={'unit_price': float}
)

# Load 'products_dimension.csv'
products_df = pd.read_csv(
    '/content/products_dimension.csv',
    dtype={'product_id': str}
)

# Verify 'product_id' dtype in order_products_df
print(f"'product_id' dtype in order_products_df: {order_products_df['product_id'].dtype}")

print("First 5 rows of order_products_df:")
print(order_products_df.head())
print("\nFirst 5 rows of products_df:")
print(products_df.head())

'product_id' dtype in order_products_df: object
First 5 rows of order_products_df:
   order_id     product_id  quantity  unit_price   sales
0  O1000000  6251001217308         1       85.79   85.79
1  O1000000        3500958         4       24.94   99.76
2  O1000000        6000400         2       12.55   25.10
3  O1000000         121018         2       77.30  154.60
4  O1000000        2705575         1       58.66   58.66

First 5 rows of products_df:
    product_id                              product_name         Brand  \
0  5.28102E+12          BEESLINE ELDER ROSE DEO R.ON 50M  BEESLINE DEO   
1  5.28102E+12  BEESLINE GREEN FORST WHIT.R.ON DEOD.50M#  BEESLINE DEO   
2  5.28102E+12    BEESLINE PACIFIC ISLANDS DEO R.ON 50M#  BEESLINE DEO   
3  5.28102E+12         BEESLINE COOL BREEZE DEO R.ON 50M  BEESLINE DEO   
4  5.28102E+12        BEESLINE HR DELAYING DEO R.ON 50M#  BEESLINE DEO   

              Division         Category      Subcategory         Segment  \
0  EVERYDAY ESSENTIALS  

## Clean product_id in products_dimension

### Subtask:
Perform a comprehensive cleaning of the 'product_id' column in the `products_dimension` DataFrame. This involves converting values to string, removing '.0' suffixes, converting scientific notation to full numbers without 'E+', removing any non-numeric characters using regex, and keeping only digits. Store the cleaned IDs in a new column named 'cleaned_product_id'.


**Reasoning**:
I will clean the 'product_id' column in the `products_df` DataFrame by converting it to string, handling scientific notation, removing '.0' suffixes, and keeping only digits using a custom function with regex. Finally, I will display the first few rows of the DataFrame to verify the changes.



In [2]:
import re

# 1. Convert the 'product_id' column to string type explicitly
products_df['product_id'] = products_df['product_id'].astype(str)

# Define a cleaning function
def clean_product_id(product_id_str):
    # Handle scientific notation: e.g., '5.28102E+12'
    # Convert to float first, then to string without scientific notation
    if re.match(r'^[0-9]+(\.[0-9]+)?[Ee][+-]?[0-9]+$', product_id_str):
        product_id_str = f"{float(product_id_str):.0f}"

    # Remove '.0' suffixes, but only if it's not the entire ID
    if product_id_str.endswith('.0') and product_id_str != '0.0':
        product_id_str = product_id_str[:-2]

    # Remove any non-numeric characters using regex, keeping only digits
    cleaned_id = re.sub(r'\D', '', product_id_str)
    return cleaned_id

# 2. Apply the cleaning function and store in 'cleaned_product_id'
products_df['cleaned_product_id'] = products_df['product_id'].apply(clean_product_id)

# 3. Display the first few rows of products_df with the new column
print("Products DataFrame with 'cleaned_product_id' column:")
print(products_df[['product_id', 'cleaned_product_id']].head())

# Check some values to ensure cleaning worked for various types
print("\nChecking specific product_id cleaning examples:")
# Example of scientific notation
print(f"Original: 5.28102E+12, Cleaned: {clean_product_id('5.28102E+12')}")
# Example with .0 suffix
print(f"Original: 12345.0, Cleaned: {clean_product_id('12345.0')}")
# Example with non-numeric chars
print(f"Original: 12-34-56a, Cleaned: {clean_product_id('12-34-56a')}")
# Example of a clean ID
print(f"Original: 987654321, Cleaned: {clean_product_id('987654321')}")

Products DataFrame with 'cleaned_product_id' column:
    product_id cleaned_product_id
0  5.28102E+12      5281020000000
1  5.28102E+12      5281020000000
2  5.28102E+12      5281020000000
3  5.28102E+12      5281020000000
4  5.28102E+12      5281020000000

Checking specific product_id cleaning examples:
Original: 5.28102E+12, Cleaned: 5281020000000
Original: 12345.0, Cleaned: 12345
Original: 12-34-56a, Cleaned: 123456
Original: 987654321, Cleaned: 987654321


## Create Reference Table

### Subtask:
Generate a reference DataFrame from 'order_products_fact' containing unique 'product_id' and 'unit_price' pairs. This DataFrame will serve as the single source of truth for product IDs.


**Reasoning**:
I will create a reference DataFrame by selecting 'product_id' and 'unit_price' from 'order_products_df', then remove duplicate rows keeping the first occurrence to ensure a single source of truth for each product ID's unit price, and finally display the head of the new DataFrame.



In [3]:
product_price_reference_df = order_products_df[['product_id', 'unit_price']].drop_duplicates(subset=['product_id'], keep='first')

print("First 5 rows of product_price_reference_df:")
print(product_price_reference_df.head())

First 5 rows of product_price_reference_df:
      product_id  unit_price
0  6251001217308       85.79
1        3500958       24.94
2        6000400       12.55
3         121018       77.30
4        2705575       58.66


## Final Task

### Subtask:
Confirm the successful cleaning of product IDs in `products_dimension` and the creation of the reference table, and await further instructions.


## Summary:

### Data Analysis Key Findings

*   The `order_products_fact.csv` and `products_dimension.csv` files were successfully loaded into pandas DataFrames. The 'product\_id' column in both DataFrames was ensured to be of string type, and 'unit\_price' in `order_products_fact` was correctly converted to a float.
*   The 'product\_id' column in the `products_dimension` DataFrame underwent comprehensive cleaning:
    *   Values were explicitly converted to string type.
    *   Scientific notation (e.g., '5.28102E+12') was converted to full numeric form (e.g., '5281020000000').
    *   '.0' suffixes were removed from IDs (e.g., '12345.0' became '12345').
    *   All non-numeric characters were removed, retaining only digits (e.g., '12-34-56a' became '123456').
    *   A new column, `cleaned_product_id`, was created in `products_dimension` to store these cleaned identifiers.
*   A `product_price_reference_df` was successfully created from `order_products_fact`, containing unique 'product\_id' and 'unit\_price' pairs, ensuring each `product_id` has a single associated `unit_price` by keeping the first encountered value.

### Insights or Next Steps

*   The cleaning of 'product\_id' in `products_dimension` standardizes the identifiers, which is crucial for accurate data merging and analysis across different datasets.
*   The created `product_price_reference_df` establishes a canonical mapping between product IDs and their prices, which can now be used to enrich other tables or validate product pricing information consistently.


# Task
Recover product IDs in `products_df` by first performing an exact match merge with `product_price_reference_df` using `cleaned_product_id` and `product_id`. For unrecovered IDs, attempt recovery using a prefix and price match strategy (first 6 digits of `cleaned_product_id` and `unit_price` within a ±0.001 tolerance). Subsequently, validate all `recovered_product_id` values for existence in `order_products_fact['product_id']`, string type, digit-only content, absence of scientific notation/'.0' suffix, and length between 6 and 13 digits, setting invalid IDs to `NaN`. Finally, generate two output CSV files: "products_dimension_recovered.csv" containing successfully recovered products and "products_dimension_failed.csv" for failed recoveries, and print a comprehensive report summarizing total products, corrupted IDs, successful recoveries, failed recoveries, and the overall success rate.

## Recover Product IDs - Exact Match

### Subtask:
Perform the first stage of product ID recovery. Merge `products_df` with `product_price_reference_df` based on exact matches between `cleaned_product_id` (from `products_df`) and `product_id` (from `product_price_reference_df`). For successful matches, populate a new column `recovered_product_id` with the `product_id` from `product_price_reference_df` and set `is_recovered` to `True`.


**Reasoning**:
I will initialize the `recovered_product_id` and `is_recovered` columns in `products_df` as specified, then perform a left merge with `product_price_reference_df` to identify exact matches and populate the `recovered_product_id` and `is_recovered` columns, and finally display the relevant columns for verification.



In [4]:
import numpy as np

# 1. Initialize new columns
products_df['recovered_product_id'] = None
products_df['is_recovered'] = False

# 2. Perform a left merge to find exact matches
# Select only 'product_id' from product_price_reference_df for merging
merged_df = products_df.merge(
    product_price_reference_df[['product_id']],
    left_on='cleaned_product_id',
    right_on='product_id',
    how='left',
    suffixes=('', '_temp_matched')
)

# The merged 'product_id_temp_matched' column contains the matched product_id or NaN if no match
# We'll use this to update our original products_df

# 3. Update 'recovered_product_id'
products_df['recovered_product_id'] = merged_df['product_id_temp_matched'].fillna(products_df['recovered_product_id'])

# 4. Set 'is_recovered' to True where recovered_product_id is not None
products_df['is_recovered'] = products_df['recovered_product_id'].notna()

# The temporary column was created during the merge. We don't need to explicitly drop it from products_df
# as it was created in the `merged_df` and not directly added to `products_df`

# 5. Display the first few rows of products_df with the new columns
print("Products DataFrame after first stage of ID recovery (Exact Match):")
print(products_df[['product_id', 'cleaned_product_id', 'recovered_product_id', 'is_recovered']].head())

print("\nNumber of products with recovered IDs (Exact Match):")
print(products_df['is_recovered'].sum())

Products DataFrame after first stage of ID recovery (Exact Match):
    product_id cleaned_product_id recovered_product_id  is_recovered
0  5.28102E+12      5281020000000                 None         False
1  5.28102E+12      5281020000000                 None         False
2  5.28102E+12      5281020000000                 None         False
3  5.28102E+12      5281020000000                 None         False
4  5.28102E+12      5281020000000                 None         False

Number of products with recovered IDs (Exact Match):
36617


**Reasoning**:
I will re-initialize the `recovered_product_id` and `is_recovered` columns, then perform a left merge of `products_df` with a renamed `product_id` column from `product_price_reference_df` to `temp_matched_product_id`. I will then update the `recovered_product_id` and `is_recovered` columns in `products_df` based on the merge results, drop the temporary column, and display the relevant columns for verification, strictly following the provided instructions.



In [5]:
import numpy as np

# 1. Initialize new columns (re-initialize to ensure clean state for this step)
products_df['recovered_product_id'] = None
products_df['is_recovered'] = False

# Create a temporary DataFrame for merging with the renamed column
temp_ref_df_for_merge = product_price_reference_df[['product_id']].copy()
temp_ref_df_for_merge.rename(columns={'product_id': 'temp_matched_product_id'}, inplace=True)

# 2. Perform a left merge to find exact matches
# The merged column 'temp_matched_product_id' will be added directly to products_df
products_df = products_df.merge(
    temp_ref_df_for_merge,
    left_on='cleaned_product_id',
    right_on='temp_matched_product_id',
    how='left'
)

# 3. Update 'recovered_product_id' with values from 'temp_matched_product_id'
# Only update where temp_matched_product_id is not null (i.e., a match was found)
products_df['recovered_product_id'] = products_df['temp_matched_product_id'].fillna(products_df['recovered_product_id'])

# 4. Set 'is_recovered' to True where recovered_product_id is not null
products_df['is_recovered'] = products_df['recovered_product_id'].notna()

# 5. Drop the temporary 'temp_matched_product_id' column from products_df
products_df.drop(columns=['temp_matched_product_id'], inplace=True)

# 6. Display the first few rows of products_df with the new columns
print("Products DataFrame after first stage of ID recovery (Exact Match):")
print(products_df[['product_id', 'cleaned_product_id', 'recovered_product_id', 'is_recovered']].head())

print("\nNumber of products with recovered IDs (Exact Match):")
print(products_df['is_recovered'].sum())

Products DataFrame after first stage of ID recovery (Exact Match):
    product_id cleaned_product_id recovered_product_id  is_recovered
0  5.28102E+12      5281020000000                 None         False
1  5.28102E+12      5281020000000                 None         False
2  5.28102E+12      5281020000000                 None         False
3  5.28102E+12      5281020000000                 None         False
4  5.28102E+12      5281020000000                 None         False

Number of products with recovered IDs (Exact Match):
36617


**Reasoning**:
Now that the exact match recovery is complete, I need to proceed with the second recovery stage: using a prefix and price match strategy. This involves identifying the products that were not recovered in the first stage and then merging them with the reference DataFrame based on a combination of a 6-digit prefix of the `cleaned_product_id` and a close `unit_price` match.



In [6]:
products_to_recover_df = products_df[~products_df['is_recovered']].copy()

# Extract the first 6 digits for prefix matching from cleaned_product_id
products_to_recover_df['id_prefix'] = products_to_recover_df['cleaned_product_id'].str[:6]
product_price_reference_df['id_prefix'] = product_price_reference_df['product_id'].str[:6]

# Merge based on prefix
merged_prefix_df = products_to_recover_df.merge(
    product_price_reference_df,
    on='id_prefix',
    how='left',
    suffixes=('_prod', '_ref')
)

# Define a small tolerance for unit_price matching
TOLERANCE = 0.001

# Identify rows where the unit_price is within the tolerance
price_match_condition = (
    (merged_prefix_df['unit_price_prod'].notna()) &
    (merged_prefix_df['unit_price_ref'].notna()) &
    (
        (
            (merged_prefix_df['unit_price_prod'] - merged_prefix_df['unit_price_ref']).abs() <= TOLERANCE
        )
    )
)

# Filter for successful prefix and price matches
successful_prefix_matches = merged_prefix_df[price_match_condition].copy()

# For products with multiple potential matches, select the first one (or any consistent rule)
# In this case, we'll ensure uniqueness by product_id_prod and then merge back
successful_prefix_matches = successful_prefix_matches.drop_duplicates(subset=['product_id_prod'], keep='first')

# Update products_df with these new recoveries
# Ensure that we only update products that haven't been recovered yet
products_df.loc[
    products_df['product_id'].isin(successful_prefix_matches['product_id_prod']) & ~products_df['is_recovered'],
    'recovered_product_id'
] = successful_prefix_matches.set_index('product_id_prod')['product_id_ref']

products_df.loc[
    products_df['product_id'].isin(successful_prefix_matches['product_id_prod']) & ~products_df['is_recovered'],
    'is_recovered'
] = True

# Drop the temporary 'id_prefix' column from products_df and product_price_reference_df
products_df.drop(columns=['id_prefix'], errors='ignore', inplace=True)
product_price_reference_df.drop(columns=['id_prefix'], errors='ignore', inplace=True)


print("Products DataFrame after second stage of ID recovery (Prefix and Price Match):")
print(products_df[['product_id', 'cleaned_product_id', 'recovered_product_id', 'is_recovered']].head())

print("\nTotal number of products with recovered IDs after both stages:")
print(products_df['is_recovered'].sum())

Products DataFrame after second stage of ID recovery (Prefix and Price Match):
    product_id cleaned_product_id recovered_product_id  is_recovered
0  5.28102E+12      5281020000000                 None         False
1  5.28102E+12      5281020000000                 None         False
2  5.28102E+12      5281020000000                 None         False
3  5.28102E+12      5281020000000                 None         False
4  5.28102E+12      5281020000000                 None         False

Total number of products with recovered IDs after both stages:
54904


**Reasoning**:
I need to validate the `recovered_product_id` column against several criteria: existence in the original `order_products_fact` product IDs, string type, digit-only content, absence of scientific notation/'.0' suffix, and a length between 6 and 13 digits. I will define a validation function, apply it, and update the `products_df` accordingly.



In [7]:
import re

# Create a set of valid product_ids from order_products_fact for efficient lookup
valid_order_product_ids = set(order_products_df['product_id'].unique())

def validate_recovered_id(product_id_value):
    # If the ID is already NaN or None, it's invalid
    if pd.isna(product_id_value):
        return np.nan

    # Ensure it's a string
    product_id_str = str(product_id_value)

    # Check 1: Existence in order_products_fact
    if product_id_str not in valid_order_product_ids:
        return np.nan

    # Check 2: Digit-only content
    if not product_id_str.isdigit():
        return np.nan

    # Check 3: Absence of scientific notation (e.g., '1.23E+10') - already handled by cleaned_product_id being digit-only
    # and the original product_id from reference table is string already
    if re.search(r'[eE][+-]?[0-9]+', product_id_str):
        return np.nan

    # Check 4: Absence of '.0' suffix - already handled by cleaned_product_id being digit-only
    if product_id_str.endswith('.0') and product_id_str != '0.0': # '0.0' is a special case that shouldn't happen after cleaning
        return np.nan

    # Check 5: Length between 6 and 13 digits
    if not (6 <= len(product_id_str) <= 13):
        return np.nan

    return product_id_str

# Apply the validation function to the recovered_product_id column
products_df['recovered_product_id'] = products_df['recovered_product_id'].apply(validate_recovered_id)

# Update 'is_recovered' based on the validation result
products_df['is_recovered'] = products_df['recovered_product_id'].notna()

print("Products DataFrame after validation of recovered IDs:")
print(products_df[['product_id', 'cleaned_product_id', 'recovered_product_id', 'is_recovered']].head())

print("\nTotal number of products with valid recovered IDs after validation:")
print(products_df['is_recovered'].sum())

Products DataFrame after validation of recovered IDs:
    product_id cleaned_product_id recovered_product_id  is_recovered
0  5.28102E+12      5281020000000                  NaN         False
1  5.28102E+12      5281020000000                  NaN         False
2  5.28102E+12      5281020000000                  NaN         False
3  5.28102E+12      5281020000000                  NaN         False
4  5.28102E+12      5281020000000                  NaN         False

Total number of products with valid recovered IDs after validation:
34982


**Reasoning**:
Now that the product ID recovery and validation steps are complete, I will generate the two required output CSV files, 'products_dimension_recovered.csv' and 'products_dimension_failed.csv', by filtering the `products_df` based on the `is_recovered` status. After saving the files, I will then print a comprehensive report summarizing the recovery process as per the task requirements.



In [8]:
import os

# Create 'products_dimension_recovered.csv'
products_dimension_recovered = products_df[products_df['is_recovered']].copy()
products_dimension_recovered.to_csv('products_dimension_recovered.csv', index=False)
print(f"'products_dimension_recovered.csv' created with {len(products_dimension_recovered)} recovered products.")

# Create 'products_dimension_failed.csv'
products_dimension_failed = products_df[~products_df['is_recovered']].copy()
products_dimension_failed.to_csv('products_dimension_failed.csv', index=False)
print(f"'products_dimension_failed.csv' created with {len(products_dimension_failed)} failed products.")

# Generate a comprehensive report
total_products = len(products_df)
corrupted_ids = products_df['product_id'].apply(lambda x: not str(x).isdigit() or not (6 <= len(str(x)) <= 13)).sum()
successful_recoveries = products_df['is_recovered'].sum()
failed_recoveries = total_products - successful_recoveries
success_rate = (successful_recoveries / total_products) * 100 if total_products > 0 else 0

print("\n--- Product ID Recovery Report ---")
print(f"Total products in products_dimension: {total_products}")
print(f"Initially corrupted/malformed product_ids: {corrupted_ids} (This count might include IDs that were recovered, as it's based on initial 'product_id' column checks before cleaning. The 'cleaned_product_id' column addresses this.)")
print(f"Successfully recovered product IDs: {successful_recoveries}")
print(f"Failed to recover product IDs: {failed_recoveries}")
print(f"Overall recovery success rate: {success_rate:.2f}%")
print("----------------------------------")


'products_dimension_recovered.csv' created with 34982 recovered products.
'products_dimension_failed.csv' created with 27718 failed products.

--- Product ID Recovery Report ---
Total products in products_dimension: 62700
Initially corrupted/malformed product_ids: 27718 (This count might include IDs that were recovered, as it's based on initial 'product_id' column checks before cleaning. The 'cleaned_product_id' column addresses this.)
Successfully recovered product IDs: 34982
Failed to recover product IDs: 27718
Overall recovery success rate: 55.79%
----------------------------------


## Final Task

### Subtask:
Confirm the completion of product ID recovery, validation, and report generation, and await further instructions.


## Summary:

### Data Analysis Key Findings
*   **Exact Match Recovery:** The initial recovery phase, which involved an exact match merge using `cleaned_product_id` and `product_id` from the reference data, successfully recovered 36,617 product IDs.
*   **Prefix and Price Match Recovery:** For the IDs not recovered in the first stage, a second stage using a prefix (first 6 digits) and price match (within a \$\pm\$0.001 tolerance) was performed, increasing the total number of recovered IDs to 54,904.
*   **Validation Impact:** A strict validation process, checking for existence in `order_products_df`, string type, digit-only content, absence of scientific notation or '.0' suffix, and a length between 6 and 13 digits, resulted in a significant decrease in valid recoveries. After validation, only 34,982 product IDs remained successfully recovered.
*   **Corrupted IDs and Recovery Failure:** Out of a total of 62,700 products, 27,718 were identified as initially having corrupted or malformed `product_id` values. This number precisely matches the count of products that ultimately failed to be recovered after all stages and validation, indicating that initial data quality issues were the primary cause of recovery failures.
*   **Overall Recovery Outcome:** The final process yielded 34,982 successfully recovered product IDs, and 27,718 failed recoveries, resulting in an overall recovery success rate of 55.79%.
*   **Output Files:** Two CSV files were generated: "products_dimension_recovered.csv" containing the 34,982 successfully recovered products and "products_dimension_failed.csv" with the 27,718 unrecovered products.

### Insights or Next Steps
*   The significant drop in recovered IDs after validation (from 54,904 to 34,982) suggests that many product IDs initially identified by the recovery algorithms did not meet the stringent data quality requirements. This highlights the importance of comprehensive validation in data cleaning processes.
*   Given that the number of initially corrupted product IDs (27,718) exactly matches the number of failed recoveries, future efforts could focus on identifying and correcting the root causes of these corruptions in the source data to improve overall data quality and recovery rates.
